### PR TITLE
Model

### PR DESCRIPTION
**Pull request by Rseiji**
1. Tokenizer code importable from `.py` script
2. Model code importable from `.py` script
3. Analysis code MVP implemented. The idea is to help processes by automating generation of analysis metrics.
4. Google colab's used notebook commited here as a temporary version of model execution code.

Local GPU adaption isn't resolved yet. Demands additional local configurations setup.

### CODE REVIEW

File: chess_commentary_model/bert_tokenizer.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    This lenght is measured by number of words. If the phrase is
+    smaller than maximun len, the remaining blank elements are filled
+    with padding tokens.
+    """
+    if default_max_len:
+      return default_max_len
+
+    max_len = 0
+    for sent in sentences:
+        input_ids = tokenizer.encode(sent, add_special_tokens=True)
```
Comment (douglasramos) [inline chess_commentary_model/bert_tokenizer.py:23]: 🚲  tokenizer e input_ids não são necessário nesse método

File: chess_commentary_model/bert_tokenizer.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+                      )
+        input_ids.append(encoded_dict['input_ids'])
+        attention_masks.append(encoded_dict['attention_mask'])
+
+    input_ids = torch.cat(input_ids, dim=0)
+    attention_masks = torch.cat(attention_masks, dim=0)
+    return (input_ids, attention_masks)
+
+
+def make_dataset(sentences, labels, max_phrase_len=None):
```
Comment (douglasramos) [inline chess_commentary_model/bert_tokenizer.py:53]: talvez um nome melhor aqui fosse tokenize_dataset

File: chess_commentary_model/bert_tokenizer.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    Returns:
+        input_ids(tensor): bert encoded phrases
+        attention_masks(tensor): tensor containing each phrase's own
+                                 attention mask
+        labels(tensor): each row's label, stored as a tensor
+    """
+    tokenizer = BertTokenizer.from_pretrained('bert-base-uncased', do_lower_case=True)
+    max_phrase_len = get_phrases_maximun_len(sentences, tokenizer, max_phrase_len)
+    input_ids, attention_masks = tokenize_sentences(sentences, max_phrase_len, tokenizer)
+    labels = torch.tensor(labels)
```
Comment (douglasramos) [inline chess_commentary_model/bert_tokenizer.py:72]: sei que talvez no python isso seja aceitável, mas acho melhor setar em outra variável. 
Ficaria algo como:
`labels_tensor_obj = torch.tensor(labels)`

File: chess_commentary_model/classification_analysis.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+    results.loc['f1_score', 'label_0'] = f1_score(labels_0, preds_0)
+    results.loc['f1_score', 'label_1'] = f1_score(labels_1, preds_1)
+    results.loc['f1_score', 'complete_dataset'] = f1_score(labels, preds)
+
+    conf_mat = confusion_matrix(df['label'], df['pred'])
+    return (results, conf_mat)
+
+
+def save_analysis(results, conf_mat, file_name_analysis, file_name_conf_mat, path=''):
```
Comment (douglasramos) [inline chess_commentary_model/classification_analysis.py:59]: 🌟

File: chess_commentary_model/classification_analysis.py [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+from sklearn.metrics import accuracy_score
+from sklearn.metrics import balanced_accuracy_score
+from sklearn.metrics import classification_report
+from sklearn.metrics import f1_score
+from sklearn.metrics import recall_score
+from sklearn.metrics import precision_score
+
+
+def compile_results(flat_predictions, flat_true_labels, test_set):
+    result_concatenation = pd.DataFrame({'pred': flat_predictions, 'labels_tmp': flat_true_labels})
```
Comment (douglasramos) [inline chess_commentary_model/classification_analysis.py:15]: por que chamamos de 'labels_tmp'?

File: chess_commentary_model/classification_analysis.py [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+from sklearn.metrics import plot_confusion_matrix
+from sklearn.metrics import accuracy_score
+from sklearn.metrics import balanced_accuracy_score
+from sklearn.metrics import classification_report
+from sklearn.metrics import f1_score
+from sklearn.metrics import recall_score
+from sklearn.metrics import precision_score
+
+
+def compile_results(flat_predictions, flat_true_labels, test_set):
```
Comment (douglasramos) [inline chess_commentary_model/classification_analysis.py:14]: Ficou um pouco confuso o que esse método está fazendo exatamente. Talvez vale um comentário explicando o que ele faz e/ou melhorar um pouco os nomes das variáveis pra dar mais semântica

File: chess_commentary_model/classification_analysis.py [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+    result_concatenation = pd.DataFrame({'pred': flat_predictions, 'labels_tmp': flat_true_labels})
+    compiled_results = pd.concat([test_set, result_concatenation], axis=1)
+
+    assert (compiled_results['label'] == compiled_results['labels_tmp']).all()
+    compiled_results = compiled_results.drop(columns=['labels_tmp'])
+    return compiled_results
+
+
+def get_test_set_analysis(df):
+    labels_0 = df[df['label'] == 0]['label']
```
Comment (douglasramos) [inline chess_commentary_model/classification_analysis.py:24]: conseguimos ser mais específicos do que usar o sufixo 0 ou 1?

File: chess_commentary_model/transformers_model.py [hunk 8]
```diff
@@ -0,0 +1,267 @@
...
+  return optimizer
+
+
+def flat_accuracy(preds, labels):
+    pred_flat = np.argmax(preds, axis=1).flatten()
+    labels_flat = labels.flatten()
+    return np.sum(pred_flat == labels_flat) / len(labels_flat)
+
+
+def format_time(elapsed):
```
Comment (douglasramos) [inline chess_commentary_model/transformers_model.py:79]: Esse método não parece ter muito a ver com esse arquivo. Talvez ele possa estar em um arquivo chamado time_utils.py

File: chess_commentary_model/transformers_model.py [hunk 9]
```diff
@@ -0,0 +1,267 @@
+##############################
```
Comment (douglasramos) [inline chess_commentary_model/transformers_model.py:1]: input_ids está em todo arquivo e eu ainda não sei o que é essa variável kkkk

**Review by douglasramos — CHANGES_REQUESTED** [review]

File: chess_commentary_model/classification_analysis.py (hunk 5 above)
Comment (Rseiji) [reply chess_commentary_model/classification_analysis.py:15]: Foi só um check extra para garantir que o ordenamento das labels nos tensores não se alterou em relação ao dataset. É portanto redundante em comparação com a coluna 'label'

File: chess_commentary_model/classification_analysis.py (hunk 7 above)
Comment (Rseiji) [reply chess_commentary_model/classification_analysis.py:24]: Acho que tá bom assim. É um problema de classificação, tá explicito o que quer dizer

File: chess_commentary_model/transformers_model.py (hunk 8 above)
Comment (Rseiji) [reply chess_commentary_model/transformers_model.py:79]: Isso aqui era pq antes o modelo cronometrava o tempo de execução. Tava dando pau e aí tirei essa funcionalidade. Futuramente, talvez seja bom implementar dnv isso (ok, em um `utils`), mas por hora vou só deletar esse método, que não está sendo usado.

File: chess_commentary_model/transformers_model.py (hunk 9 above)
Comment (Rseiji) [reply chess_commentary_model/transformers_model.py:1]: troquei por tokenized_sentences. Aparentemente se trata de um argumento do objeto que faz a tokenização, do qual extraímos as sentenças tokenizadas pelo bert.

File: chess_commentary_model/bert_tokenizer.py [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+        max_len = max(max_len, len(input_ids))
+  
+    print('Max sentence length: ', max_len)
+    return max_len
+
+
+def tokenize_sentences(sentences, max_length, tokenizer):
+    """Tokenize all of the sentences and map the tokens to their word IDs.
+    """
+    input_ids = []
```
[the remark this answers is not in the document]
Comment (Rseiji) [reply chess_commentary_model/bert_tokenizer.py]: ver resposta ao ultimo comentario desse code review

File: chess_commentary_model/transformers_model.py (hunk 9 above)
Comment (douglasramos) [reply chess_commentary_model/transformers_model.py:1]: boa!

File: chess_commentary_model/transformers_model.py (hunk 8 above)
Comment (douglasramos) [reply chess_commentary_model/transformers_model.py]: beleza!

File: chess_commentary_model/classification_analysis.py (hunk 7 above)
Comment (douglasramos) [reply chess_commentary_model/classification_analysis.py:26]: boa, verdade, não tinha avaliado isso

File: chess_commentary_model/classification_analysis.py (hunk 5 above)
Comment (douglasramos) [reply chess_commentary_model/classification_analysis.py:17]: beleza

**Review by douglasramos — APPROVED** [review]
LGTM!!
